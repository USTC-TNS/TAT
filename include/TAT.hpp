/* TAT.hpp
 * Copyright (C) 2019  Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef TAT_HPP_
#define TAT_HPP_

#ifndef TAT_VERSION
#define TAT_VERSION "unknown"
#endif // TAT_VERSION

#include <cassert>
#include <cstring>
#include <cmath>
#include <iostream>
#include <fstream>
#include <map>
#include <set>
#include <vector>
#include <algorithm>
#include <memory>
#include <functional>
#include <complex>

#define PASS std::cerr << "calling a passing function at " << __FILE__ << ":" << __LINE__ << " in " << __PRETTY_FUNCTION__ << std::endl, exit(233)
#define ENABLE_IF(...) class = typename std::enable_if<__VA_ARGS__::value>::type

#if (!defined TAT_USE_CPU && !defined TAT_USE_CUDA && !defined TAT_USE_DCU && !defined TAT_USE_SW)
#if !defined TAT_DEFAULT
#warning use CPU by default
#endif
#define TAT_USE_CPU
#endif

#ifdef TAT_EXTREME
#warning EXTREME compile may cost much of compile time
#endif // TAT_EXTREME

#ifdef TAT_USE_CPU
extern "C"
{
#define MKL_Complex8 std::complex<float>
#define MKL_Complex16 std::complex<double>
#include <mkl.h>
} // extern "C"
#include <hptt.h>
#ifdef TAT_EXTREME
#include <../src/hptt.cpp>
#include <../src/plan.cpp>
#include <../src/transpose.cpp>
#include <../src/utils.cpp>
#endif // TAT_EXTREME
#include <rang.hpp>

// SVD
#if (defined TAT_USE_GESDD && defined TAT_USE_GESVD) || (defined TAT_USE_GESVD && defined TAT_USE_GESVDX) || (defined TAT_USE_GESVDX && defined TAT_USE_GESDD)
#error only one of GESDD, GESVD and GESVDX could be in use
#endif
#if (!defined TAT_USE_GESDD && !defined TAT_USE_GESVD && !defined TAT_USE_GESVDX)
#if !defined TAT_DEFAULT
#warning must use one of GESDD, GESVD and GESVDX, default use GESVD now
#endif
#define TAT_USE_GESVD
#endif

// QR
#if (defined TAT_USE_GEQRF && defined TAT_USE_GEQP3)
#error only one of GEQRF and GEQP3 could be in use
#endif
#if (!defined TAT_USE_GEQRF && !defined TAT_USE_GEQP3)
#if !defined TAT_DEFAULT
#warning must use one of GEQRF and GEQP3, default use GEQRF now
#endif
#define TAT_USE_GEQRF
#endif

#ifdef TAT_USE_GEQP3
#error GEQP3 is current unusable
#endif

#endif // TAT_USE_CPU

namespace TAT {
  template<class T>
  class is_scalar {
   public:
    static constexpr bool value = std::is_scalar<T>::value || std::is_same<T, std::complex<float>>::value || std::is_same<T, std::complex<double>>::value || std::is_same<T, std::complex<long double>>::value;
  };

  template<class Base>
  class RealBaseClass {
   public:
    using type=Base;
  };
  template<>
  class RealBaseClass<std::complex<float>> {
   public:
    using type=float;
  };
  template<>
  class RealBaseClass<std::complex<double>> {
   public:
    using type=double;
  };
  template<>
  class RealBaseClass<std::complex<long double>> {
   public:
    using type=long double;
  };
  template<class T>
  using RealBase = typename RealBaseClass<T>::type;

  enum class Device : unsigned char {CPU, CUDA, DCU, SW};

  namespace legs {
    class Legs {
     public:
      unsigned char id;
      bool operator==(const Legs& other) const {
        return id==other.id;
      }
      bool operator!=(const Legs& other) const {
        return id!=other.id;
      }
      bool operator<(const Legs& other) const {
        return id<other.id;
      }
      Legs() = default;
      Legs(const std::string& name) {
        try {
          id = name2id.at(name);
        } catch (const std::out_of_range& e) {
          id = total++;
          name2id[name]=id;
          id2name[id]=name;
        } // exsit name
      }
      static unsigned char total;
      static std::map<std::string, unsigned char> name2id;
      static std::map<unsigned char, std::string> id2name;

      friend std::ostream& operator<<(std::ostream& out, const Legs& value) {
        try {
          return out << id2name.at(value.id);
        } catch (const std::out_of_range& e) {
          return out << "UserDefinedLeg" << value.id;
        } // get name
      } // operator<<
    }; // class Legs

    unsigned char Legs::total = 0;
    std::map<std::string, unsigned char> Legs::name2id = {};
    std::map<unsigned char, std::string> Legs::id2name = {};
  } // namespace legs
  using legs::Legs;

  namespace legs_name {
#define TAT_DefineLeg(x) static const TAT::Legs x(#x)
#define TAT_DefineLegs(n) \
      TAT_DefineLeg(Phy##n); TAT_DefineLeg(Left##n); TAT_DefineLeg(Right##n); TAT_DefineLeg(Up##n); TAT_DefineLeg(Down##n); \
      TAT_DefineLeg(LeftUp##n); TAT_DefineLeg(LeftDown##n); TAT_DefineLeg(RightUp##n); TAT_DefineLeg(RightDown##n)
#define TAT_Legs \
      TAT_DefineLegs(); TAT_DefineLegs(1); TAT_DefineLegs(2); TAT_DefineLegs(3); TAT_DefineLegs(4); \
      TAT_DefineLegs(5); TAT_DefineLegs(6); TAT_DefineLegs(7); TAT_DefineLegs(8); TAT_DefineLegs(9)
    TAT_Legs;
#undef TAT_Legs
#undef TAT_DefineLegs
#define TAT_DefineLegs(n) \
      TAT_DefineLeg(Tmp##n##0); TAT_DefineLeg(Tmp##n##1); TAT_DefineLeg(Tmp##n##2); TAT_DefineLeg(Tmp##n##3); TAT_DefineLeg(Tmp##n##4); \
      TAT_DefineLeg(Tmp##n##5); TAT_DefineLeg(Tmp##n##6); TAT_DefineLeg(Tmp##n##7); TAT_DefineLeg(Tmp##n##8); TAT_DefineLeg(Tmp##n##9)
#define TAT_Legs \
      TAT_DefineLegs(); TAT_DefineLegs(1); TAT_DefineLegs(2); TAT_DefineLegs(3); TAT_DefineLegs(4); \
      TAT_DefineLegs(5); TAT_DefineLegs(6); TAT_DefineLegs(7); TAT_DefineLegs(8); TAT_DefineLegs(9)
    TAT_Legs;
#undef TAT_Legs
#undef TAT_DefineLegs
#undef TAT_DefineLeg
    // total = 190 currently
  } // namespace legs_name

  using Size = std::size_t;
  using Rank = unsigned int;

  namespace data {
    template<Device device, class Base>
    class Magic;
#define DefineData(x) \
      namespace x { \
        template<class Base> \
        class Data; \
      } \
      template<class Base> \
      class Magic<Device::x, Base>{ \
       public: \
        using type=x::Data<Base>; \
      }

    DefineData(CPU);
    DefineData(CUDA);
    DefineData(DCU);
    DefineData(SW);
#undef DefineData

    template<Device device=Device::CPU, class Base=double, ENABLE_IF(is_scalar<Base>)>
    using Data = typename Magic<device, Base>::type;
  } // namespace data
  using data::Data;

  namespace block {
    template<Device device=Device::CPU, class Base=double>
    class Block;
  } // namespace block
  using block::Block;

  namespace tensor {
    template<Device device=Device::CPU, class Base=double>
    class Tensor;
  } // namespace tensor
  //using tensor::Tensor;
  template<Device device=Device::CPU, class Base=double>
  using Tensor=Block<device, Base>;

  namespace node {
    template<Device device=Device::CPU, class Base=double>
    class Node;
  } // namespace node
  using node::Node;

  namespace lazy {
    template<Device device=Device::CPU, class Base=double>
    class Lazy;
  } // namespace lazy
  using lazy::Lazy;

  namespace site {
    template<Device device=Device::CPU, class Base=double>
    class Site;
  } // namespace site
  using site::Site;

  namespace lattice {
    template<int dimension>
    class Dimension;

    template<class Tags=Dimension<2>, Device device=Device::CPU, class Base=double>
    class Lattice;
  } // namespace lattice
  using site::Site;
} // namespace TAT

#include "TAT/Data.hpp"
#include "TAT/Block.hpp"
// Tensor
#include "TAT/Node.hpp"
#include "TAT/Lazy.hpp"

#include "TAT/Site.hpp"
#include "TAT/Lattice.hpp"

#endif // TAT_HPP_
