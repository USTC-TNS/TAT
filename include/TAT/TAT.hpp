/**
 * \file TAT.hpp
 *
 * Copyright (C) 2019-2020 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#pragma once
#ifndef TAT_HPP
#define TAT_HPP

#ifndef __cplusplus
#error only work for c++
#endif

#ifdef _MSVC_LANG
#if _MSVC_LANG < 201703L
#error require c++17 or later
#endif
#else
#if __cplusplus < 201703L
#error require c++17 or later
#endif
#endif

#ifdef _WIN32
#include <windows.h>
#endif

// 开关说明
// TAT_USE_MPI 定义以开启MPI支持, cmake可对此进行定义
// TAT_USE_SIMPLE_NAME 定义以使用原始字符串作为name

namespace TAT {
   /**
    * \brief TAT的版本号
    */
   inline const char* version = "0.0.5";

   /**
    * \brief Debug模式中, 将在程序末尾打印一行友情提示, 过早的优化是万恶之源, 同时控制windows下终端的色彩模式
    */
   struct evil_t {
      evil_t() {
#ifdef _WIN32
         HANDLE output_handle = GetStdHandle(STD_OUTPUT_HANDLE);
         DWORD output_mode = 0;
         GetConsoleMode(output_handle, &output_mode);
         SetConsoleMode(output_handle, output_mode | ENABLE_VIRTUAL_TERMINAL_PROCESSING);
         HANDLE error_handle = GetStdHandle(STD_ERROR_HANDLE);
         DWORD error_mode = 0;
         GetConsoleMode(error_handle, &error_mode);
         SetConsoleMode(error_handle, error_mode | ENABLE_VIRTUAL_TERMINAL_PROCESSING);
#endif
      }
      ~evil_t();
   };
   inline const evil_t evil;
} // namespace TAT

// clang-format off
#include "tensor.hpp"
#include "implement.hpp"
#include "tools.hpp"
// clang-format on

#endif
