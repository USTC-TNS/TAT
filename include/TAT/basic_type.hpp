/**
 * \file basic_type.hpp
 *
 * Copyright (C) 2019-2020 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#pragma once
#ifndef TAT_BASIC_TYPE_HPP
#define TAT_BASIC_TYPE_HPP

#include <complex>
#include <type_traits>

namespace TAT {
   /**
    * \brief 张量的秩的类型
    */
   using Rank = unsigned short;
   /**
    * \brief 张量数据维度大小和数据本身大小的类型
    */
   using Size = unsigned long;

   /**
    * \brief Z2对称性的类型
    */
   using Z2 = bool;
   /**
    * \brief U1对称性的类型
    */
   using U1 = int;
   /**
    * \brief 费米子数目的类型
    */
   using Fermi = short;

   /**
    * \brief 费米箭头方向的类型, false和true分别表示出入
    */
   using Arrow = bool;

   /**
    * \brief 判断一个类型是否是标量类型, 修复了std::scalar不能判断std::complex的问题
    * \tparam T 如果T是标量类型, 则value为true
    * \see is_scalar_v
    */
   template<typename T>
   struct is_scalar : std::is_scalar<T> {};
   template<typename T>
   struct is_scalar<std::complex<T>> : std::is_scalar<T> {};
   template<typename T>
   constexpr bool is_scalar_v = is_scalar<T>::value;

   /**
    * \brief 取对应的实数类型, 在svd, norm等地方会用到
    * \tparam T 如果T是std::complex<S>, 则type为S, 若T为其他标量类型, 则type为T本身, 否则为void
    * \see real_base_t
    */
   template<typename T>
   struct real_base : std::conditional<is_scalar<T>::value, T, void> {};
   template<typename T>
   struct real_base<std::complex<T>> : std::conditional<is_scalar<T>::value, T, void> {};
   template<typename T>
   using real_base_t = typename real_base<T>::type;

   /**
    * \brief 判断是否是复数类型
    * \see is_complex_v
    */
   template<typename T>
   struct is_complex : std::is_same<T, std::complex<real_base_t<T>>> {};
   template<typename T>
   constexpr bool is_complex_v = is_complex<T>::value;
   /**
    * \brief 判断是否是实数类型
    * \see is_real_v
    */
   template<typename T>
   struct is_real : std::is_same<T, real_base_t<T>> {};
   template<typename T>
   constexpr bool is_real_v = is_real<T>::value;
} // namespace TAT
#endif
