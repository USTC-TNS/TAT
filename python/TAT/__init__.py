import cppyy
cppyy.load_library("blas")
cppyy.load_library("lapack")
cppyy.include("../include/TAT/TAT.hpp")
std = cppyy.gbl.std
TAT = cppyy.gbl.TAT
