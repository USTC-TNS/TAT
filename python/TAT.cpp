#include <sstream>

#include <pybind11/functional.h>
#include <pybind11/operators.h>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

#define TAT_ALWAYS_COLOR
#include "TAT/TAT.hpp"

namespace TAT {
   namespace py = pybind11;

   template<class Type, class Args>
   auto implicit_init() {
      py::implicitly_convertible<Args, Type>();
      return py::init<Args>();
   }

   template<class Type, class Args, class Func>
   auto implicit_init(Func&& func) {
      py::implicitly_convertible<Args, Type>();
      return py::init(func);
   }

   template<class ScalarType, class Symmetry>
   auto declare_tensor(py::module& m, const char* name) {
      using T = Tensor<ScalarType, Symmetry>;
      using E = Edge<Symmetry>;
      return py::class_<T>(m, name)
            .def(py::self + py::self)
            .def(ScalarType() + py::self)
            .def(py::self + ScalarType())
            .def(py::self += py::self)
            .def(py::self += ScalarType())
            .def(py::self - py::self)
            .def(ScalarType() - py::self)
            .def(py::self - ScalarType())
            .def(py::self -= py::self)
            .def(py::self -= ScalarType())
            .def(py::self * py::self)
            .def(ScalarType() * py::self)
            .def(py::self * ScalarType())
            .def(py::self *= py::self)
            .def(py::self *= ScalarType())
            .def(py::self / py::self)
            .def(ScalarType() / py::self)
            .def(py::self / ScalarType())
            .def(py::self /= py::self)
            .def(py::self /= ScalarType())
            .def("__str__",
                 [](const T& tensor) {
                    auto out = std::stringstream();
                    out << tensor;
                    return out.str();
                 })
            .def("__repr__",
                 [](const T& tensor) {
                    auto out = std::stringstream();
                    out << "Tensor";
                    out << tensor;
                    return out.str();
                 })
            .def("shape",
                 [](const T& tensor) {
                    auto out = std::stringstream();
                    out << '{' << console_green << "names" << console_origin << ':';
                    out << tensor.names;
                    out << ',' << console_green << "edges" << console_origin << ':';
                    out << tensor.core->edges;
                    out << '}';
                    return out.str();
                 })
            .def(py::init<const std::vector<Name>&, const std::vector<E>&>())
            .def(implicit_init<T, std::pair<std::vector<Name>, std::vector<E>>>([](const std::pair<std::vector<Name>, std::vector<E>>& p) {
               return std::make_unique<T>(p.first, p.second);
            }))
            // TODO problem: `TAT.TensorDoubleU1Symmetry(list("AB"),[{0:2},{0:2}])` not work
            // https://github.com/pybind/pybind11/issues/2182
            .def(implicit_init<T, ScalarType>())
            .def("value", [](const T& tensor) -> ScalarType { return tensor; })
            .def("copy", &T::copy)
            .def("same_shape", &T::same_shape)
            .def("map", [](const T& tensor, std::function<ScalarType(ScalarType)>& function) { return tensor.map(function); })
            .def("transform", [](T& tensor, std::function<ScalarType(ScalarType)>& function) -> T& { return tensor.transform(function); })
            .def("set", [](T& tensor, std::function<ScalarType()>& function) -> T& { return tensor.set(function); })
            .def("zero", [](T& tensor) -> T& { return tensor.zero(); })
            .def(
                  "test",
                  [](T& tensor, ScalarType first, ScalarType step) -> T& { return tensor.test(first, step); },
                  py::arg("first") = 0,
                  py::arg("step") = 1)
            // TODO block and at
            // TODO to
            .def("norm_max", &T::template norm<-1>)
            .def("norm_num", &T::template norm<0>)
            .def("norm_1", &T::template norm<1>)
            .def("norm_2", &T::template norm<2>);
      // TODO edge contract svd
   }

   template<class Symmetry>
   auto declare_edge(py::module& m, const char* name) {
      auto result = py::class_<Edge<Symmetry>>(m, name)
                          .def_readwrite("map", &Edge<Symmetry>::map)
                          .def(implicit_init<Edge<Symmetry>, Size>())
                          .def(implicit_init<Edge<Symmetry>, std::map<Symmetry, Size>>())
                          .def(implicit_init<Edge<Symmetry>, const std::set<Symmetry>&>())
                          .def("__str__",
                               [](const Edge<Symmetry>& edge) {
                                  auto out = std::stringstream();
                                  out << edge;
                                  return out.str();
                               })
                          .def("__repr__", [](const Edge<Symmetry>& edge) {
                             auto out = std::stringstream();
                             out << "Edge";
                             out << edge;
                             return out.str();
                          });
      if constexpr (is_fermi_symmetry_v<Symmetry>) {
         result = result.def_readwrite("arrow", &Edge<Symmetry>::arrow).def(py::init<Arrow, std::map<Symmetry, Size>>());
      }
      return result;
   }

   PYBIND11_MODULE(TAT, m) {
      m.doc() = "TAT is A Tensor library!";
      // name
      py::class_<Name>(m, "Name")
            .def(implicit_init<Name, const std::string&>())
            .def("__repr__", [](const Name& name) { return "Name[" + id_to_name.at(name.id) + "]"; })
            .def("__str__", [](const Name& name) { return id_to_name.at(name.id); })
            .def_readonly("id", &Name::id)
            .def_property_readonly("name", [](const Name& name) { return id_to_name.at(name.id); });

      // symmetry
      m.attr("Symmetry") = py::dict(
            py::arg("NoSymmetry") = py::class_<NoSymmetry>(m, "NoSymmetry")
                                          .def(py::init<>())
                                          .def("__repr__", [=](const NoSymmetry& symmetry) { return std::string("NoSymmetry") + "[" + +"]"; })
                                          .def("__str__",
                                               [=](const NoSymmetry& symmetry) {
                                                  auto out = std::stringstream();
                                                  out << symmetry;
                                                  return out.str();
                                               }),
            py::arg("Z2Symmetry") =
                  py::class_<Z2Symmetry>(m, "Z2Symmetry")
                        .def(implicit_init<Z2Symmetry, Z2>())
                        .def("__repr__",
                             [=](const Z2Symmetry& symmetry) { return std::string("Z2Symmetry") + "[" + std::to_string(symmetry.z2) + "]"; })
                        .def("__str__",
                             [=](const Z2Symmetry& symmetry) {
                                auto out = std::stringstream();
                                out << symmetry;
                                return out.str();
                             })
                        .def_readwrite("z2", &Z2Symmetry::z2),
            py::arg("U1Symmetry") =
                  py::class_<U1Symmetry>(m, "U1Symmetry")
                        .def(implicit_init<U1Symmetry, U1>())
                        .def("__repr__",
                             [=](const U1Symmetry& symmetry) { return std::string("U1Symmetry") + "[" + std::to_string(symmetry.u1) + "]"; })
                        .def("__str__",
                             [=](const U1Symmetry& symmetry) {
                                auto out = std::stringstream();
                                out << symmetry;
                                return out.str();
                             })
                        .def_readwrite("u1", &U1Symmetry::u1),
            py::arg("FermiSymmmetry") =
                  py::class_<FermiSymmetry>(m, "FermiSymmetry")
                        .def(implicit_init<FermiSymmetry, Fermi>())
                        .def("__repr__",
                             [=](const FermiSymmetry& symmetry) { return std::string("FermiSymmetry") + "[" + std::to_string(symmetry.fermi) + "]"; })
                        .def("__str__",
                             [=](const FermiSymmetry& symmetry) {
                                auto out = std::stringstream();
                                out << symmetry;
                                return out.str();
                             })
                        .def_readwrite("fermi", &FermiSymmetry::fermi),
            py::arg("FermiZ2Symmmetry") = py::class_<FermiZ2Symmetry>(m, "FermiZ2Symmetry")
                                                .def(py::init<Fermi, Z2>())
                                                .def(implicit_init<FermiZ2Symmetry, const std::pair<Fermi, Z2>&>([](const std::pair<Fermi, Z2>& p) {
                                                   return std::make_unique<FermiZ2Symmetry>(p.first, p.second);
                                                }))
                                                .def("__repr__",
                                                     [=](const FermiZ2Symmetry& symmetry) {
                                                        return std::string("FermiZ2Symmetry") + "[" + std::to_string(symmetry.fermi) + "," +
                                                               std::to_string(symmetry.z2) + "]";
                                                     })
                                                .def("__str__",
                                                     [=](const FermiZ2Symmetry& symmetry) {
                                                        auto out = std::stringstream();
                                                        out << symmetry;
                                                        return out.str();
                                                     })
                                                .def_readwrite("fermi", &FermiZ2Symmetry::fermi)
                                                .def_readwrite("z2", &FermiZ2Symmetry::z2),
            py::arg("FermiU1Symmmetry") = py::class_<FermiU1Symmetry>(m, "FermiU1Symmetry")
                                                .def(py::init<Fermi, U1>())
                                                .def(implicit_init<FermiU1Symmetry, const std::pair<Fermi, U1>&>([](const std::pair<Fermi, U1>& p) {
                                                   return std::make_unique<FermiU1Symmetry>(p.first, p.second);
                                                }))
                                                .def("__repr__",
                                                     [=](const FermiU1Symmetry& symmetry) {
                                                        return std::string("FermiU1Symmetry") + "[" + std::to_string(symmetry.fermi) + "," +
                                                               std::to_string(symmetry.u1) + "]";
                                                     })
                                                .def("__str__",
                                                     [=](const FermiU1Symmetry& symmetry) {
                                                        auto out = std::stringstream();
                                                        out << symmetry;
                                                        return out.str();
                                                     })
                                                .def_readwrite("fermi", &FermiU1Symmetry::fermi)
                                                .def_readwrite("u1", &FermiU1Symmetry::u1));
      // edge
      m.attr("Edge") = py::dict(
#define DECLARE_EDGE(S) py::arg(#S) = declare_edge<S>(m, "Edge" #S)
            DECLARE_EDGE(NoSymmetry),
            DECLARE_EDGE(Z2Symmetry),
            DECLARE_EDGE(U1Symmetry),
            DECLARE_EDGE(FermiSymmetry),
            DECLARE_EDGE(FermiZ2Symmetry),
            DECLARE_EDGE(FermiU1Symmetry)
#undef DECLARE_EDGE
      );
      // tensor
      declare_tensor<double, NoSymmetry>(m, "TensorDoubleNoSymmetry");
      declare_tensor<double, U1Symmetry>(m, "TensorDoubleU1Symmetry");
      // TODO: python support
   }
} // namespace TAT
