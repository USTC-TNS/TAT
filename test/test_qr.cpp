/**
 * Copyright (C) 2020-2021 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <TAT/TAT.hpp>

#include "run_test.hpp"

using Tensor = typename TAT::Tensor<float, TAT::NoSymmetry>;

void run_test() {
   auto a = Tensor({"A", "B"}, {5, 10}).range();
   std::cout << a << '\n';
   {
      auto [q, r] = a.qr('r', {"A"}, "newQ", "newR");
      std::cout << q << '\n'; // "B" "newQ"
      std::cout << r << '\n'; // "A" "newR"
      std::cout << (q.contract(r, {{"newQ", "newR"}}) - a).norm<-1>() << '\n';
   }
   {
      auto [q, r] = a.qr('r', {"B"}, "newQ", "newR");
      std::cout << q << '\n';
      std::cout << r << '\n';
      std::cout << (q.contract(r, {{"newQ", "newR"}}) - a).norm<-1>() << '\n';
   }
   auto b = Tensor({"A", "B"}, {10, 5}).range();
   std::cout << b << '\n';
   {
      auto [q, r] = b.qr('r', {"A"}, "newQ", "newR");
      std::cout << q << '\n';
      std::cout << r << '\n';
      std::cout << (q.contract(r, {{"newQ", "newR"}}) - b).norm<-1>() << '\n';
   }
   {
      auto [q, r] = b.qr('r', {"B"}, "newQ", "newR");
      std::cout << q << '\n';
      std::cout << r << '\n';
      std::cout << (q.contract(r, {{"newQ", "newR"}}) - b).norm<-1>() << '\n';
   }
}
