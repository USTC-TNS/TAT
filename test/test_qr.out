{[32mnames[0m:[A,B],[32medges[0m:[5,10],[32mblocks[0m:[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49]}
{[32mnames[0m:[newQ,B],[32medges[0m:[5,10],[32mblocks[0m:[0,-0.0592349,-0.11847,-0.177705,-0.23694,-0.296174,-0.355409,-0.414644,-0.473879,-0.533114,-0.587754,-0.494951,-0.402147,-0.309344,-0.216541,-0.123738,-0.0309344,0.0618688,0.154672,0.247475,0.0641468,-0.293392,-0.44628,0.312486,0.325564,0.0580972,0.226223,0.113803,0.25238,-0.613027,0.685856,0.0473901,-0.548653,-0.250472,-0.180274,-0.143368,-0.0641434,0.233206,-0.00755273,0.228013,0.410482,-0.787615,0.402897,-0.0171625,-0.0423003,0.0745458,0.052995,-0.175403,-0.00558217,0.0871433]}
{[32mnames[0m:[A,newR],[32medges[0m:[5,5],[32mblocks[0m:[-16.8819,0,0,0,0,-43.5376,-17.0139,0,0,0,-70.1933,-34.0278,-8.4984e-07,0,0,-96.849,-51.0418,-1.74643e-07,-5.68982e-06,0,-123.505,-68.0557,-3.93532e-06,-1.03686e-05,-1.02069e-05]}
1.52588e-05
{[32mnames[0m:[A,newQ],[32medges[0m:[5,5],[32mblocks[0m:[0,-0.774597,0.494152,0.274102,-0.284044,-0.182574,-0.516397,-0.392145,-0.738956,0.0129251,-0.365148,-0.258199,-0.139899,0.359011,0.80718,-0.547723,9.37478e-07,-0.520382,0.402437,-0.51696,-0.730297,0.258198,0.558272,-0.296594,-0.0191009]}
{[32mnames[0m:[newR,B],[32medges[0m:[5,10],[32mblocks[0m:[-54.7723,-56.598,-58.4237,-60.2495,-62.0752,-63.901,-65.7267,-67.5524,-69.3782,-71.2039,0,-1.291,-2.58199,-3.87298,-5.16398,-6.45498,-7.74597,-9.03696,-10.328,-11.6189,0,0,-2.15706e-06,-4.41644e-06,-5.1045e-06,-5.10846e-06,-9.62818e-06,-1.32667e-05,-1.39061e-05,-1.43799e-05,0,0,0,-9.79609e-08,4.83972e-07,1.81624e-06,2.91697e-07,-1.9391e-07,8.46687e-07,1.4338e-06,0,0,0,0,-1.09175e-06,-2.16389e-06,-1.07816e-06,-2.1273e-08,-1.23493e-06,-2.2181e-06]}
5.72205e-06
{[32mnames[0m:[A,B],[32medges[0m:[10,5],[32mblocks[0m:[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49]}
{[32mnames[0m:[newQ,B],[32medges[0m:[5,5],[32mblocks[0m:[0,-0.182574,-0.365148,-0.547723,-0.730297,-0.774597,-0.516398,-0.258199,4.20509e-08,0.258199,0.619988,-0.727957,-0.171492,0.0469013,0.232559,-0.120533,-0.405967,0.843581,0.0128714,-0.329952,0.0329566,0.0725924,0.242821,-0.835245,0.486876]}
{[32mnames[0m:[A,newR],[32medges[0m:[10,5],[32mblocks[0m:[-5.47723,0,0,0,0,-14.6059,-6.45497,0,0,0,-23.7346,-12.9099,2.81745e-06,0,0,-32.8634,-19.3649,3.78992e-06,6.12119e-07,0,-41.9921,-25.8199,5.2328e-06,4.31292e-09,9.2973e-08,-51.1208,-32.2749,7.42259e-06,1.10431e-06,4.07737e-07,-60.2495,-38.7298,7.48138e-06,1.43422e-06,5.53083e-08,-69.3782,-45.1848,1.08596e-05,-1.26043e-07,1.37403e-06,-78.5069,-51.6398,4.36559e-06,6.81822e-07,-3.81758e-07,-87.6356,-58.0948,1.65327e-05,4.06373e-07,-3.71494e-07]}
1.14441e-05
{[32mnames[0m:[A,newQ],[32medges[0m:[10,5],[32mblocks[0m:[0,-0.587754,0.729228,-0.0701056,0.343051,-0.0592349,-0.494951,-0.26696,0.803843,-0.113891,-0.11847,-0.402147,-0.204626,-0.359469,-0.342421,-0.177705,-0.309344,-0.171427,-0.362003,-0.232979,-0.23694,-0.216541,-0.15461,-0.190791,-0.0623001,-0.296174,-0.123737,-0.137793,-0.0985332,0.0597585,-0.355409,-0.0309341,-0.120976,0.0726792,0.196495,-0.414644,0.0618692,-0.104159,0.00702906,0.345768,-0.473879,0.154673,-0.0709606,0.00448912,0.409954,-0.533114,0.247474,0.502284,0.19286,-0.603434]}
{[32mnames[0m:[newR,B],[32medges[0m:[5,5],[32mblocks[0m:[-84.4097,-87.0753,-89.7409,-92.4064,-95.072,0,-1.70139,-3.40278,-5.10417,-6.80557,0,0,-4.51954e-06,-8.96363e-06,-8.02488e-06,0,0,0,-3.77464e-07,-4.8969e-07,0,0,0,0,2.63408e-06]}
9.29832e-06
